File: limo/limo_learnning/src/limo_action_server.cpp
//导入头文件
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include <geometry_msgs/msg/twist.hpp>
#include <std_msgs/msg/bool.hpp>
#include <limo_msgs/action/limo_action.hpp>

#include <iostream>

using namespace std;
using namespace std::placeholders;

// 定义一个LimoActionServer类，继承自rclcpp::Node
class LimoActionServer : public rclcpp::Node
{

  public:
  // 简化名称
  using limo_action = limo_msgs::action::LimoAction;
  using GoalHandleLimo = rclcpp_action::ServerGoalHandle<limo_action>;
  // 构造函数，初始化节点名称为"limo_action_server"
  explicit LimoActionServer(const rclcpp::NodeOptions & server_options = rclcpp::NodeOptions())
  : Node("limo_action_server",server_options)
  {
    // 创建发布者
    pub_vel = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel",10);
    // 创建动作服务器
    this->limo_action_server_ = rclcpp_action::create_server<limo_action>(
      this->get_node_base_interface(),
      this->get_node_clock_interface(),
      this->get_node_logging_interface(),
      this->get_node_waitables_interface(),
      "limo_action",
      std::bind(&LimoActionServer::action_goal,this,_1,_2),
      std::bind(&LimoActionServer::action_cancel,this,_1),
      std::bind(&LimoActionServer::action_accepted,this,_1));
  }
  private:

  geometry_msgs::msg::Twist vel_cmd;
  // 定义发布者指针
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_vel;
  // 定义action 服务器指针
  rclcpp_action::Server<limo_action>::SharedPtr limo_action_server_;

  // 处理目标请求的回调函数
  rclcpp_action::GoalResponse action_goal(
      const rclcpp_action::GoalUUID &uuid,std::shared_ptr<const limo_action::Goal> limo_goal
  )
  {
    (void)uuid;

    RCLCPP_INFO(rclcpp::get_logger("server"), "Got goal request with order X: %f", limo_goal->x);
    RCLCPP_INFO(rclcpp::get_logger("server"), "Got goal request with order Z: %f", limo_goal->z);
    // 限制目标值的大小
    if (limo_goal->x > 1.5 || limo_goal->z > 1.5) 
    {
      // 拒绝请求
      RCLCPP_INFO(rclcpp::get_logger("server"),"REJECT");
      return rclcpp_action::GoalResponse::REJECT;

    }
    // 接受并执行请求
    RCLCPP_INFO(rclcpp::get_logger("server"),"GoalResponse");
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;

  }
  //用于处理取消目标请求
  rclcpp_action::CancelResponse action_cancel(
      const std::shared_ptr<GoalHandleLimo> limo_cancel
  )
  {
      RCLCPP_INFO(rclcpp::get_logger("server"), "Got request to cancel goal");
      // 忽略参数  
      (void)limo_cancel;
      // 接受请求
      return rclcpp_action::CancelResponse::ACCEPT;
      RCLCPP_INFO(rclcpp::get_logger("server"),"action_cancel");

  }
  // 执行目标请求的回调函数
  void execute(
    const std::shared_ptr<GoalHandleLimo> limo_execute)
  {

    RCLCPP_INFO(rclcpp::get_logger("server"), "Executing goal");
    rclcpp::Rate loop_rate(1);
    // 获取目标
    const auto goal = limo_execute->get_goal();
    // 创建反馈指针
    auto feedback = std::make_shared<limo_action::Feedback>();
    // 获取状态
    auto & sequence = feedback->status;
    // 创建结果指针
    auto result = std::make_shared<limo_action::Result>();
    // 目标信息
    vel_cmd.linear.x = goal.get()->x;
    vel_cmd.linear.y = goal.get()->y;
    vel_cmd.angular.z = goal.get()->z;

    // 检查是否处于执行状态
    if (limo_execute->is_active()) {
      limo_execute->succeed(result);
      pub_vel->publish(vel_cmd);
      limo_execute->publish_feedback(feedback);
      RCLCPP_INFO(rclcpp::get_logger("server"), "Goal Succeeded");
      rclcpp::sleep_for(std::chrono::seconds(2));
      result->success = true; 

      vel_cmd.linear.x = 0.0;
      vel_cmd.linear.y = 0.0;
      vel_cmd.angular.z = 0.0;
      pub_vel->publish(vel_cmd);
    }


  }
  // 处理接受请求的回调函数
  void action_accepted(
      const std::shared_ptr<GoalHandleLimo> limo_accepted
  )
  {
    // 创建一个线程来执行目标请求
    std::thread{std::bind(&LimoActionServer::execute,this,_1),limo_accepted}.detach();

  }

};




int main(int argc, char const *argv[])
{
    rclcpp::init(argc, argv);
    //运行action 服务节点
    rclcpp::spin(std::make_shared<LimoActionServer>());   

    rclcpp::shutdown();

    return 0;
}

File: limo/limo_learnning_py/limo_learnning_py/limo_action_server.py
import rclpy  # 导入ROS 2的Python客户端库
from rclpy.node import Node  # 导入Node类以创建ROS 2节点
from rclpy.action import ActionServer, CancelResponse, GoalResponse  # 导入ActionServer类和相关响应类
from geometry_msgs.msg import Twist  # 导入Twist消息类型，用于发布速度命令
from limo_msgs.action import LimoAction  # 导入LimoAction消息类型
import threading  # 导入线程库，用于并发处理
import time  # 导入时间库，用于延时处理

# 定义LimoActionServer类，继承自Node
class LimoActionServer(Node):

    def __init__(self):  # 构造函数
        super().__init__('limo_action_server')  # 初始化节点，命名为'limo_action_server'
        self.pub_vel = self.create_publisher(Twist, 'cmd_vel', 10)  # 创建速度命令发布者，发布到'cmd_vel'话题
        self.limo_action_server = ActionServer(  # 创建动作服务器
            self,  # 当前节点
            LimoAction,  # 动作类型
            'limo_action',  # 动作名称
            execute_callback=self.execute_callback,  # 执行目标的回调函数
        )

    def execute_callback(self, goal_handle):  # 执行目标的回调函数
        self.get_logger().info('Executing goal')  # 记录正在执行目标的信息
        vel_cmd = Twist()  # 创建Twist消息实例
        feedback_msg = LimoAction.Feedback()  # 创建反馈消息实例
        result = LimoAction.Result()  # 创建结果消息实例

        vel_cmd.linear.x = goal_handle.request.x  # 设置线速度X值
        vel_cmd.linear.y = goal_handle.request.y  # 设置线速度Y值
        vel_cmd.angular.z = goal_handle.request.z  # 设置角速度Z值

        if goal_handle.is_active:  # 如果目标句柄处于活动状态
            self.pub_vel.publish(vel_cmd)  # 发布速度命令
            self.get_logger().info('Goal Succeeded')  # 记录目标成功的信息
            goal_handle.succeed()  # 标记目标成功
            result.success = True  # 设置结果成功

            feedback_msg.status = 1  # 设置反馈消息状态
            goal_handle.publish_feedback(feedback_msg)  # 发布反馈消息

            time.sleep(2)  # 延时2秒

            vel_cmd.linear.x = 0.0  # 停止线速度X
            vel_cmd.linear.y = 0.0  # 停止线速度Y
            vel_cmd.angular.z = 0.0  # 停止角速度Z
            self.pub_vel.publish(vel_cmd)  # 发布停止命令
            result.success = True
        return result  # 返回结果消息

def main(args=None):  # 主函数
    rclpy.init(args=args)  # 初始化rclpy
    action_server = LimoActionServer()  # 创建LimoActionServer实例
    rclpy.spin(action_server)  # 保持节点运行，等待和处理ROS 2消息
    rclpy.shutdown()  # 关闭rclpy

if __name__ == '__main__':  # 检查是否为主程序入口
    main()  # 调用主函数
